crates/types/src/macros.rs

```diff
@@ -0,0 +1,11 @@
+/// Utility for removing boilerplate from From impls
+#[macro_export]
+macro_rules! impl_from {
+    ($($t1:ty => $t2:ty, | $i:ident | {$e:expr},)*) => {$(
+        impl From<$t1> for $t2 {
+            fn from($i: $t1) -> Self {
+                $e
+            }
+        }
+    )*};
+}
```

Comment:
I like this macro a lot, I pulled it from @neonphog's code. I'm barely using it in this PR but we can start using it.